components/cluster/command/exec.go

```diff
@@ -46,87 +31,12 @@ func newExecCmd() *cobra.Command {
 			clusterName := args[0]
 			teleCommand = append(teleCommand, scrubClusterName(clusterName))
 
-			exist, err := tidbSpec.Exist(clusterName)
-			if err != nil {
-				return perrs.AddStack(err)
-			}
-
-			if !exist {
-				return perrs.Errorf("cannot execute command on non-exists cluster %s", clusterName)
-			}
-
-			logger.EnableAuditLog()
-			metadata, err := spec.ClusterMetadata(clusterName)
-			if err != nil && !errors.Is(perrs.Cause(err), meta.ErrValidate) {
-				return err
-			}
-
-			filterRoles := set.NewStringSet(gOpt.Roles...)
-			filterNodes := set.NewStringSet(gOpt.Nodes...)
-
-			var shellTasks []task.Task
-			uniqueHosts := map[string]int{} // host -> ssh-port
-			metadata.Topology.IterInstance(func(inst spec.Instance) {
-				if _, found := uniqueHosts[inst.GetHost()]; !found {
-					if len(gOpt.Roles) > 0 && !filterRoles.Exist(inst.Role()) {
-						return
-					}
-
-					if len(gOpt.Nodes) > 0 && !filterNodes.Exist(inst.GetHost()) {
-						return
-					}
-
-					uniqueHosts[inst.GetHost()] = inst.GetSSHPort()
-				}
-			})
-
-			for host := range uniqueHosts {
-				shellTasks = append(shellTasks,
-					task.NewBuilder().
-						Shell(host, opt.command, opt.sudo).
-						Build())
-			}
-
-			t := task.NewBuilder().
-				SSHKeySet(
-					spec.ClusterPath(clusterName, "ssh", "id_rsa"),
-					spec.ClusterPath(clusterName, "ssh", "id_rsa.pub")).
-				ClusterSSH(metadata.Topology, metadata.User, gOpt.SSHTimeout).
-				Parallel(shellTasks...).
-				Build()
-
-			execCtx := task.NewContext()
-			if err := t.Execute(execCtx); err != nil {
-				if errorx.Cast(err) != nil {
-					// FIXME: Map possible task errors and give suggestions.
-					return err
-				}
-				return perrs.Trace(err)
-			}
-
-			// print outputs
-			for host := range uniqueHosts {
-				stdout, stderr, ok := execCtx.GetOutputs(host)
-				if !ok {
-					continue
-				}
-				log.Infof("Outputs of %s on %s:",
-					color.CyanString(opt.command),
-					color.CyanString(host))
-				if len(stdout) > 0 {
-					log.Infof("%s:\n%s", color.GreenString("stdout"), stdout)
-				}
-				if len(stderr) > 0 {
-					log.Infof("%s:\n%s", color.RedString("stderr"), stderr)
-				}
-			}
-
-			return nil
+			return deployer.Exec(clusterName, opt, gOpt)
```

Comment:
Also, `deployer` is not a proper name for operations other than `deploy` and `scale-out`.